### PR TITLE
[Frontend] Refactor prompt processing

### PR DESCRIPTION
**Pull request by DarkLight1337**
This PR refactors various parts of the OpenAI-compatible server:

- The `_validate_prompt_and_tokenize` method has been decomposed so that `prompt` and `prompt_ids` are processed separately.
- The logging of `prompt` and `prompt_ids` has been moved from `vllm.AsyncLLMEngine` to `vllm.entrypoints.logger.RequestLogger` such that redundant data is no longer passed into the core engine. This also enables logging for tokenization endpoints.
- Explicitly handle LoRA and prompt adapters for embedding endpoints. Previously, they were silently ignored.
- Use a different prefix for `request_id` based on the endpoint type:
  - Completion API: `cmpl-*` (as before)
  - Chat Completion API: `chat-*`
  - Embeddings API: `embd-*`
  - Tokenization API: `tokn-*`
- Fixed various type errors

### CODE REVIEW

**Comment by DarkLight1337** [comment]
Seems that #4032 alone fixed the flaky tests. Perhaps we can merge this first then?

**Comment by DarkLight1337** [comment]
Seems that #4032 fixed the LoRA bugs, however `entrypoints-test` is still failing.

**Comment by DarkLight1337** [comment]
**Update:** I found that it is due to a bug in my refactored parsing, my bad. I have fixed it just now.

**Comment by DarkLight1337** [comment]
@njhill The tests now pass. To move on with #3978, perhaps we can merge this PR first?

**Comment by DarkLight1337** [comment]
> **Update:** I found that it is due to a bug in my refactored parsing, my bad. I have fixed it just now.

I'm updating `test_batch_completions` to better catch this error. Previously, the error only occurred in `test_echo_logprob_completion` which misled me from the real cause.

**Comment by njhill** [comment]
@DarkLight1337 sorry for the hold-up, I will hopefully get to this tomorrow including looking at reconciling with #3512.

**Comment by DarkLight1337** [comment]
No worries. Do note that I have ported some of the changes (mostly related to type annotations) to #4355, so perhaps we can wait until that has been merged first.

**Comment by DarkLight1337** [comment]
Alright, #4355 has been merged so I have updated this branch accordingly.

**Comment by DarkLight1337** [comment]
Heads-up: I have copied `OpenAIServing._parse_prompt_input_or_inputs` to `vllm.inputs.parse_and_batch_prompt` in #4328. If that other PR is merged first, I'll make the change in this one accordingly.

**Comment by njhill** [comment]
@DarkLight1337 sorry, going to look at this next, I did a review of #4328 at least :)

**Comment by DarkLight1337** [comment]
I have merged #4328 into this PR in advance, so you might see additional diffs. These diffs should disappear once #4328 is merged into `main`.

**Comment by DarkLight1337** [comment]
@njhill any update?

**Comment by DarkLight1337** [comment]
@njhill I have updated this PR based on your #6227. PTAL.

File: vllm/entrypoints/openai/serving_tokenization.py
Comment (njhill) [inline vllm/entrypoints/openai/serving_tokenization.py:131]: I think it may be better to allow this - you could consider a "model" based on a prompt adapter to just have the same tokenizer.

**Review by njhill — COMMENTED** [review]
Thanks @DarkLight1337, looks great. Sorry for taking an age to finish the other PR and get to this.

My only concern is whether some folks might make use of the logging in the engine with different front ends.

File: vllm/entrypoints/openai/serving_tokenization.py
Comment (njhill) [inline vllm/entrypoints/openai/serving_tokenization.py:82]: I think this should be
```suggestion
            add_special_tokens = True
```

**Comment by njhill** [comment]
Looks like [this](https://github.com/vllm-project/vllm/blob/1689219ebf01c750de492271832e27c39df38648/tests/entrypoints/openai/test_tokenization.py#L68) line also needs to be removed from the tokenization test.

**Comment by DarkLight1337** [comment]
> Looks like [this](https://github.com/vllm-project/vllm/blob/1689219ebf01c750de492271832e27c39df38648/tests/entrypoints/openai/test_tokenization.py#L68) line also needs to be removed from the tokenization test.

Actually, let me add `add_special_tokens` to Completions API as well to make things consistent. It's a vLLM feature originally introduced by #5278.

File: vllm/entrypoints/openai/serving_tokenization.py
Comment (DarkLight1337) [reply vllm/entrypoints/openai/serving_tokenization.py]: See [comment](https://github.com/vllm-project/vllm/pull/4028#issuecomment-2237882353)

File: vllm/entrypoints/openai/cli_args.py
Comment (DarkLight1337) [reply vllm/entrypoints/openai/cli_args.py:139]: It's just to keep the same ordering of arguments as before. Since the last argument for `AsyncEngineArgs` is `--disable-log-requests`, I think it would be more appropriate to display `--max-log-len` after it.

File: vllm/entrypoints/openai/serving_tokenization.py
Comment (DarkLight1337) [reply vllm/entrypoints/openai/serving_tokenization.py:128]: Should we log a warning or just silently ignore this?

**Comment by DarkLight1337** [comment]
> My only concern is whether some folks might make use of the logging in the engine with different front ends.

I've moved out the logging to a separate class `vllm.entrypoints.logger.RequestLogger`, so it should be easy for them to manually log the requests if need be.

File: vllm/entrypoints/openai/serving_embedding.py
Comment (DarkLight1337) [reply vllm/entrypoints/openai/serving_embedding.py]: Fixed!

File: vllm/entrypoints/openai/serving_tokenization.py
Comment (njhill) [reply vllm/entrypoints/openai/serving_tokenization.py:128]: I would say silently ignore, it's not really an error.

File: vllm/entrypoints/openai/api_server.py
Comment (njhill) [reply vllm/entrypoints/openai/api_server.py]: OK here though I was just suggesting some deduplication in how the various OpenAIServingXX instances are constructed

File: vllm/entrypoints/openai/protocol.py
Comment (njhill) [inline vllm/entrypoints/openai/protocol.py:178]: Why remove the `Optional` here but not for all the other `bool` fields?

File: vllm/entrypoints/openai/protocol.py
Comment (njhill) [inline vllm/entrypoints/openai/protocol.py:175]: These classes define the HTTP API JSON scheme so I think `true` is actually more appropriate?

File: vllm/entrypoints/openai/protocol.py
Comment (DarkLight1337) [reply vllm/entrypoints/openai/protocol.py:178]: Let me update the rest as well.

**Comment by DarkLight1337** [comment]
I have finished addressing your comments.

**Review by njhill — APPROVED** [review]
Thanks @DarkLight1337!

**Pull request merged**
